File: TO2/Parser/SuffixOperation.cs
using System.Collections.Generic;
using KontrolSystem.Parsing;
using KontrolSystem.TO2.AST;

namespace KontrolSystem.TO2.Parser;

public interface ISuffixOperation {
    Expression GetExpression(Expression target, Position start, Position end);
}

public interface IAssignSuffixOperation : ISuffixOperation {
    Expression AssignExpression(Expression target, Operator op, Expression value, Position start, Position end);
}

public readonly struct IndexGetSuffix(IndexSpec indexSpec) : ISuffixOperation, IAssignSuffixOperation {
    public readonly IndexSpec indexSpec = indexSpec;

    public Expression GetExpression(Expression target, Position start, Position end) {
        return new IndexGet(target, indexSpec, start, end);
    }

    public Expression AssignExpression(Expression target, Operator op, Expression value, Position start, Position end) {
        return new IndexAssign(target, indexSpec, op, value, start, end);
    }
}

public readonly struct FieldGetSuffix(string fieldName) : ISuffixOperation, IAssignSuffixOperation {
    public readonly string fieldName = fieldName;

    public Expression GetExpression(Expression target, Position start, Position end) {
        return new FieldGet(target, fieldName, start, end);
    }

    public Expression AssignExpression(Expression target, Operator op, Expression value, Position start,
        Position end) {
        return new FieldAssign(target, fieldName, op, value, start, end);
    }
}

public readonly struct MethodCallSuffix(string methodName, List<Expression> arguments) : ISuffixOperation {
    public readonly string methodName = methodName;
    public readonly List<Expression> arguments = arguments;

    public Expression GetExpression(Expression target, Position start, Position end) {
        return new MethodCall(target, methodName, arguments, start, end);
    }
}

public readonly struct OperatorSuffix(Operator op) : ISuffixOperation {
    public readonly Operator op = op;

    public Expression GetExpression(Expression target, Position start, Position end) {
        return new UnarySuffix(target, op, start, end);
    }
}

File: Tools/vscode/to2-syntax/server/src/to2/suffix-operation.ts
import { Expression } from "./ast";
import { FieldGet } from "./ast/field-get";
import { IndexGet } from "./ast/index-get";
import { IndexSpec } from "./ast/index-spec";
import { MethodCall } from "./ast/method-call";
import { Operator } from "./ast/operator";
import { UnarySuffix } from "./ast/unary-suffix";
import { IndexAssign } from "./ast/index-assign";
import { FieldAssign } from "./ast/field-assign";
import { InputPosition, WithPosition } from "../parser";

export interface SuffixOperation {
  getExpression(
    target: Expression,
    start: InputPosition,
    end: InputPosition
  ): Expression;
}

export interface AssignSuffixOperation extends SuffixOperation {
  assignExpression(
    target: Expression,
    op: Operator,
    value: Expression,
    start: InputPosition,
    end: InputPosition
  ): Expression;
}

export class IndexGetSuffix implements SuffixOperation, AssignSuffixOperation {
  constructor(public readonly indexSpec: IndexSpec) {}
  getExpression(
    target: Expression,
    start: InputPosition,
    end: InputPosition
  ): Expression {
    return new IndexGet(target, this.indexSpec, start, end);
  }

  assignExpression(
    target: Expression,
    op: Operator,
    value: Expression,
    start: InputPosition,
    end: InputPosition
  ): Expression {
    return new IndexAssign(target, this.indexSpec, op, value, start, end);
  }
}

export class FieldGetSuffix implements SuffixOperation, AssignSuffixOperation {
  constructor(public readonly fieldName: WithPosition<string>) {}
  getExpression(
    target: Expression,
    start: InputPosition,
    end: InputPosition
  ): Expression {
    return new FieldGet(target, this.fieldName, start, end);
  }

  assignExpression(
    target: Expression,
    op: Operator,
    value: Expression,
    start: InputPosition,
    end: InputPosition
  ): Expression {
    return new FieldAssign(target, this.fieldName, op, value, start, end);
  }
}

export class MethodCallSuffix implements SuffixOperation {
  constructor(
    public readonly methodName: WithPosition<string>,
    public readonly args: Expression[]
  ) {}

  getExpression(
    target: Expression,
    start: InputPosition,
    end: InputPosition
  ): Expression {
    return new MethodCall(target, this.methodName, this.args, start, end);
  }
}

export class OperatorSuffix implements SuffixOperation {
  constructor(public readonly op: WithPosition<Operator>) {}

  getExpression(
    target: Expression,
    start: InputPosition,
    end: InputPosition
  ): Expression {
    return new UnarySuffix(target, this.op, start, end);
  }
}
